compiler/x/codegen/OMRTreeEvaluator.cpp

```diff
@@ -1767,6 +1767,367 @@ static void arrayCopy16BitPrimitive(TR::Node* node, TR::Register* dstReg, TR::Re
+   generateLabelInstruction(TR::InstOpCode::label, node, copy4ORMoreBytesLabel, cg);
+   generateRegImmInstruction(TR::InstOpCode::CMPRegImm4(), node, sizeReg, 8, cg);
+   generateLabelInstruction(TR::InstOpCode::JA4, node, copy10ORMoreBytesLabel, cg);
+
+   // 4, 6, 8 Bytes
+   generateRegMemInstruction(TR::InstOpCode::L4RegMem, node, tmpReg1, generateX86MemoryReference(srcReg, sizeReg, 0, -4, cg), cg);
+   generateRegMemInstruction(TR::InstOpCode::L4RegMem, node, tmpReg2, generateX86MemoryReference(srcReg, 0, cg), cg);
+   generateMemRegInstruction(TR::InstOpCode::S4MemReg, node, generateX86MemoryReference(dstReg, sizeReg, 0, -4, cg), tmpReg1, cg);
+   generateMemRegInstruction(TR::InstOpCode::S4MemReg, node, generateX86MemoryReference(dstReg, 0, cg), tmpReg2, cg);
+   generateLabelInstruction(TR::InstOpCode::JMP4, node, mainEndLabel, cg);
+
+   // ---------------------------------
+   generateLabelInstruction(TR::InstOpCode::label, node, copy10ORMoreBytesLabel, cg);
+
+   // 10-16 Bytes
+   generateRegMemInstruction(TR::InstOpCode::L8RegMem, node, tmpReg1, generateX86MemoryReference(srcReg, sizeReg, 0, -8, cg), cg);
+   generateRegMemInstruction(TR::InstOpCode::L8RegMem, node, tmpReg2, generateX86MemoryReference(srcReg, 0, cg), cg);
+   generateMemRegInstruction(TR::InstOpCode::S8MemReg, node, generateX86MemoryReference(dstReg, sizeReg, 0, -8, cg), tmpReg1, cg);
+   generateMemRegInstruction(TR::InstOpCode::S8MemReg, node, generateX86MemoryReference(dstReg, 0, cg), tmpReg2, cg);
+   generateLabelInstruction(TR::InstOpCode::JMP4, node, mainEndLabel, cg);
+
+   // ---------------------------------
+   generateLabelInstruction(TR::InstOpCode::label, node, copy18ORMoreBytesLabel, cg);
+   generateRegImmInstruction(TR::InstOpCode::CMPRegImm4(), node, sizeReg, 32, cg);
+   generateLabelInstruction(TR::InstOpCode::JA4, node, copy34ORMoreBytesLabel, cg);
+
+   // 18-32 Bytes
+   generateRegMemInstruction(TR::InstOpCode::MOVDQURegMem, node, tmpXmmYmmReg1, generateX86MemoryReference(srcReg, sizeReg, 0, -16, cg), cg);
+   generateRegMemInstruction(TR::InstOpCode::MOVDQURegMem, node, tmpXmmYmmReg2, generateX86MemoryReference(srcReg, 0, cg), cg);
+   generateMemRegInstruction(TR::InstOpCode::MOVDQUMemReg, node, generateX86MemoryReference(dstReg, sizeReg, 0, -16, cg), tmpXmmYmmReg1, cg);
+   generateMemRegInstruction(TR::InstOpCode::MOVDQUMemReg, node, generateX86MemoryReference(dstReg, 0, cg), tmpXmmYmmReg2, cg);
+   generateLabelInstruction(TR::InstOpCode::JMP4, node, mainEndLabel, cg);
+
+   // ---------------------------------
+   generateLabelInstruction(TR::InstOpCode::label, node, copy34ORMoreBytesLabel, cg);
+   generateRegImmInstruction(TR::InstOpCode::CMPRegImm4(), node, sizeReg, 64, cg);
+   generateLabelInstruction(TR::InstOpCode::JA4, node, repMovsLabel, cg);
+
+   // 34-64 Bytes
+   generateRegMemInstruction(TR::InstOpCode::InstOpCode::VMOVDQUYmmMem, node, tmpXmmYmmReg1, generateX86MemoryReference(srcReg, sizeReg, 0, -32, cg), cg);
```

Comment:
@BradleyWood : can you confirm if the use of these YMM instructions requires a check for AVX2?  This logic is only guarded with an AVX check at the moment.

Reply:
That instruction only requires AVX.

Why not use AVX-512 to do the 64-byte load/store in a single instruction if available. Otherwise, fall back two 32-byte copies.
```
if (cg->comp()->target().cpu.supportsFeature(OMR_FEATURE_X86_AVX512F)) {
   // vmovdqu zmm0, [src]
   // vmovdqu [dest], zmm0
   generateRegMemInstruction(TR::InstOpCode::MOVDQURegMem, node, tmpXmmYmmReg2, generateX86MemoryReference(srcReg, 0, cg), cg, OMR::X86::EVEX_L512);
   generateMemRegInstruction(TR::InstOpCode::MOVDQUMemReg, node, generateX86MemoryReference(dstReg, 0, cg), tmpXmmYmmReg2, cg, OMR::X86::EVEX_L512);
} else {
   // vmovdqu ymm0, -32[src]
   // vmovdqu ymm1,   0[src]
   // vmovdqu -32[dest], ymm0
   // vmovdqu   0[dest], ymm1
   // ...
   // Use OMR::X86::VEX_l256 Encoding; assumes AVX is supported
}
```

Reply:
Using 64-byte load/store in a single instruction is only applicable when the copy size is exactly 64 bytes. It will add an extra test when `copy size == 64` or `copy size > 64`. 

```
copy34ORMoreBytesLabel: 
   if size > 64 => repMovsLabel
   ...
repMovsLabel:
    use rep movsd for size > 64
```
==>

```
copy34ORMoreBytesLabel: 
   if size >= 64 => copy64ORMoreBytesLabel
   ...
copy64ORMoreBytesLabel:
   if size > 64 => repMovsLabel
   // 64-byte load/store in a single instruction
   ...
repMovsLabel:
    use rep movsd for size > 64
```

Reply:
Am I missing something, because this compiles to two 32-byte loads, and two 32-byte stores = 64 bytes?
```
vmovdqu	ymm0, ymmword ptr [rsi+1*rcx-0x20]		# VMOVDQUYmmMem, SymRef [#453 -32]
vmovdqu	ymm1, ymmword ptr [rsi]		# VMOVDQUYmmMem
vmovdqu	ymmword ptr [rdi+1*rcx-0x20], ymm0		# VMOVDQUMemYmm, SymRef [#455 -32]
vmovdqu	ymmword ptr [rdi], ymm1		# VMOVDQUMemYmm
```

Reply:
This code is used for copy size ranging from 34 to 64 bytes. It is not just for copy size as exactly 64 bytes.  For example, if we need to copy 48 bytes, it does: 

- load 32 bytes from [src + size - 32] = [src + 48 - 32] = [src + 16]  into temp1
- load 32 bytes from [src] into temp2
- store temp1 into [dst + size - 32]
- store temp2 into [dst]

```
0 1 2 ......15 16............ 31 32 33.......47
               | --------- temp1 ------------|
|----------- temp2 ------------|
```

Reply:
I experimented using 64-byte load/store in a single instruction to copy 64 bytes. The performance is worse compared to the current implementation. As I suspect previously, I think the impact is from the extra test on the copy size. Therefore, it makes sense to  keep the current implementation for copying size 34 to 64 bytes as it is.

(1) current implementation 
(2) check if the copy size is 64 bytes, if so, using load/store in a single instruction

Copy Size | (1) ns/ops | (2) ns/ops
---|---|---
62 Bytes | 135 | 135
64 Bytes | 135 | 174
64 Bytes | 593 | 673